crates/vm2-interface/src/lib.rs

```diff
@@ -91,3 +91,40 @@ pub use self::{state_interface::*, tracer_interface::*};
 
 mod state_interface;
 mod tracer_interface;
+
+/// Returned from [`Tracer::after_instruction`] to indicate whether the VM should continue execution.
+#[derive(Debug)]
+pub enum ExecutionStatus {
+    /// Continue execution.
+    Running,
+    /// Stop or suspend execution.
+    Stopped(ExecutionEnd),
```

Comment:
I don't quite like that a tracer can stop execution with `ProgramFinished` or `Reverted` outcomes; it looks weird / unnatural. Not quite sure about `Panicked` either. Maybe, it'd be sufficient to just allow tracers suspend execution on a hook?

Reply:
I completely disagree on this one. When the validation tracer ends execution prematurely, that is certainly a panic. When the validation tracer ends execution because the validation step ended, that is ProgramFinished. (I currently don't do the latter in the tracer but I would have if I did the hooks via tracer)

Reply:
In my mind, all "normal" ways to end execution have certain implicit invariants associated with them, such as having only the root call frame (e.g., unwinding in case of panics). Having these invariants breakable by tracers looks like a suboptimal design decision; hooks are good in this regard because they explicitly possible to encounter at any point in execution and allow to the host to decide how to proceed. A downside of this approach is that the host needs to be aware of hooks that can be set by tracers, but specifically for validation it's not an issue (it's easy to agree that a certain hook has "execution stopped by tracer" semantics).

Reply:
One advantage that I can see is that the VM can know from the hook number which tracer is responsible for the hook. I can also see that stopping while there are stack frames is unusual and dangerous.

On the other hand I have thought that suspending should not be allowed because it is completely useless for (external) tracers, as the multivm will just continue running immediately.

Reply:
I think the only use for this feature is "I want to stop the VM early to save on computation". The validation tracer would also like to signal if it stopped because of success or failure but that can be done via the tracer.

Maybe there should be another VM end state: StoppedByTracer?

Reply:
From the correctness perspective, an additional add state would be optimal. If it doesn't regress performance, I'd go with it.

Reply:
I will implement that